File: src/main/java/com/library/users/Utilisateur.java
package com.library.users;

import com.library.observers.NotificationObserver;
import java.util.ArrayList;
import java.util.List;

/**
 * Classe abstraite Utilisateur
 * Définit le comportement partagé pour tous les types d'utilisateurs
 */
public abstract class Utilisateur implements NotificationObserver {
    private String id;
    private String nom;
    private String email;
    private List<String> livresEmpruntés;
    
    // Limite d'emprunts par type d'utilisateur
    protected abstract int getMaxEmprunt();

    public Utilisateur(String id, String nom, String email) {
        this.id = id;
        this.nom = nom;
        this.email = email;
        this.livresEmpruntés = new ArrayList<>();
    }

    public String getId() {
        return id;
    }

    public String getNom() {
        return nom;
    }

    public String getEmail() {
        return email;
    }

    public List<String> getLivresEmpruntés() {
        return livresEmpruntés;
    }

    public void addLivreEmpruntu(String livreTitre) {
        if (livresEmpruntés.size() < getMaxEmprunt()) {
            livresEmpruntés.add(livreTitre);
        }
    }

    public void removeLivreEmpruntu(String livreTitre) {
        livresEmpruntés.remove(livreTitre);
    }

    public int getNombreLivresEmpruntés() {
        return livresEmpruntés.size();
    }

    public boolean peutEmprunter() {
        return livresEmpruntés.size() < getMaxEmprunt();
    }

    @Override
    public void update(String message) {
        System.out.println("[Notification pour " + nom + "] " + message);
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "{" +
                "id='" + id + '\'' +
                ", nom='" + nom + '\'' +
                ", email='" + email + '\'' +
                ", livresEmpruntés=" + livresEmpruntés.size() +
                '}';
    }
}

File: js/src/users/Utilisateur.js
const NotificationObserver = require('../observers/NotificationObserver');

/**
 * Classe abstraite Utilisateur
 * Définit le comportement partagé pour tous les types d'utilisateurs
 */
class Utilisateur extends NotificationObserver {
    constructor(id, nom, email) {
        super();
        this.id = id;
        this.nom = nom;
        this.email = email;
        this.livresEmpruntés = [];
    }

    /**
     * Méthode abstraite - doit être implémentée par les sous-classes
     * @returns {number} Limite d'emprunts
     */
    getMaxEmprunt() {
        throw new Error('getMaxEmprunt() doit être implémenté');
    }

    getId() {
        return this.id;
    }

    getNom() {
        return this.nom;
    }

    getEmail() {
        return this.email;
    }

    getLivresEmpruntés() {
        return this.livresEmpruntés;
    }

    /**
     * Ajouter un livre emprunté
     * @param {string} livreTitre - Le titre du livre
     */
    addLivreEmpruntu(livreTitre) {
        if (this.livresEmpruntés.length < this.getMaxEmprunt()) {
            this.livresEmpruntés.push(livreTitre);
        }
    }

    /**
     * Retirer un livre emprunté
     * @param {string} livreTitre - Le titre du livre
     */
    removeLivreEmpruntu(livreTitre) {
        const index = this.livresEmpruntés.indexOf(livreTitre);
        if (index > -1) {
            this.livresEmpruntés.splice(index, 1);
        }
    }

    getNombreLivresEmpruntés() {
        return this.livresEmpruntés.length;
    }

    /**
     * Vérifier si l'utilisateur peut emprunter
     * @returns {boolean} True si limite non atteinte
     */
    peutEmprunter() {
        return this.livresEmpruntés.length < this.getMaxEmprunt();
    }

    /**
     * Implémenter l'interface NotificationObserver
     * @param {string} message - Le message de notification
     */
    update(message) {
        console.log(`[Notification pour ${this.nom}] ${message}`);
    }

    toString() {
        return `${this.constructor.name}{
            id: '${this.id}',
            nom: '${this.nom}',
            email: '${this.email}',
            livresEmpruntés: ${this.livresEmpruntés.length}
        }`;
    }
}

module.exports = Utilisateur;
